lib/core/cursor.js

```diff
@@ -601,12 +609,124 @@ Cursor.prototype._initializeCursor = function(callback) {
+      !cursor.cmd.tailable &&
+      !cursor.cmd.awaitData
+    ) {
+      return setCursorNotified(cursor, callback);
+    }
+
+    callback(err, result);
+  }
+
+  const queryCallback = (err, r) => {
+    if (err) {
+      return done(err);
+    }
+
+    const result = r.message;
+    if (result.queryFailure) {
+      return done(new MongoError(result.documents[0]), null);
+    }
+
+    // Check if we have a command cursor
+    if (
+      Array.isArray(result.documents) &&
+      result.documents.length === 1 &&
+      (!cursor.cmd.find || (cursor.cmd.find && cursor.cmd.virtual === false)) &&
+      (typeof result.documents[0].cursor !== 'string' ||
+        result.documents[0]['$err'] ||
+        result.documents[0]['errmsg'] ||
+        Array.isArray(result.documents[0].result))
+    ) {
+      // We have an error document, return the error
+      if (result.documents[0]['$err'] || result.documents[0]['errmsg']) {
+        return done(new MongoError(result.documents[0]), null);
+      }
+
+      // We have a cursor document
+      if (result.documents[0].cursor != null && typeof result.documents[0].cursor !== 'string') {
+        var id = result.documents[0].cursor.id;
+        // If we have a namespace change set the new namespace for getmores
+        if (result.documents[0].cursor.ns) {
+          cursor.ns = result.documents[0].cursor.ns;
+        }
+        // Promote id to long if needed
+        cursor.cursorState.cursorId = typeof id === 'number' ? Long.fromNumber(id) : id;
+        cursor.cursorState.lastCursorId = cursor.cursorState.cursorId;
+        cursor.cursorState.operationTime = result.documents[0].operationTime;
+
+        // If we have a firstBatch set it
+        if (Array.isArray(result.documents[0].cursor.firstBatch)) {
+          cursor.cursorState.documents = result.documents[0].cursor.firstBatch; //.reverse();
+        }
+
+        // Return after processing command cursor
+        return done(null, result);
+      }
+
+      if (Array.isArray(result.documents[0].result)) {
+        cursor.cursorState.documents = result.documents[0].result;
+        cursor.cursorState.cursorId = Long.ZERO;
+        return done(null, result);
+      }
+    }
+
+    // Otherwise fall back to regular find path
+    const cursorId = result.cursorId || 0;
+    cursor.cursorState.cursorId = cursorId instanceof Long ? cursorId : Long.fromNumber(cursorId);
+    cursor.cursorState.documents = result.documents;
+    cursor.cursorState.lastCursorId = result.cursorId;
+
+    // Transform the results with passed in transformation method if provided
+    if (
+      cursor.cursorState.transforms &&
+      typeof cursor.cursorState.transforms.query === 'function'
+    ) {
+      cursor.cursorState.documents = cursor.cursorState.transforms.query(result);
+    }
+
+    done(null, result);
+  };
+
+  if (cursor.operation) {
+    executeOperation(cursor.topology, cursor.operation, (err, result, server) => {
```

Comment:
is this a nested `executeOperation` call?

Reply:
only insofar as methods like `toArray` and `next`, etc use `executeOperation`. I agree with you that we should disambiguate the operation execution from `promiseOrCallback` very soon.